### PR TITLE
Add unit_aliases context manager

### LINKED ISSUE #11246: Context manager for handling "bad" units
Lots of existing data has unit strings which are problematic in some way: plural rather than singular (e.g. "Angstrom" vs "Angstroms"), alternate spellings (e.g. "Angstrom" vs "Angstroem" or "meter" vs "metre", although the latter should probably be allowed), or alternate shortened forms (e.g. "Ang" for "Angstrom"). The options for handing this are not ideal: users could create custom units to allow parsing the strings, though this would cause these problematic strings to propagate; try to do something with the ` UnrecognizedUnit` to correct the units (though this does tend to produce a large number of warnings, which is not ideal); or based on prior experience/knowledge hard code the units (which makes their code less robust to changes/new inputs).

An idea I had would be to have a context manager that allows setting temporary aliases—this avoid the need to define new units, and reduce warning fatigue, and prevent the propagation of bad unit strings. Usage would be something like:
```python
import astropy.units as u
from some_project import DataClass

with u.unit_aliases({"Angstroms": u.AA}):
    data = DataClass.read("some_file.fits")
```

### PR DESCRIPTION
**Pull request by aragilar**
Closes #11246.

I'm not sure the test file I picked is the best (it's one of the hosted astropy ones, a local one would likely be better), so if people have suggestions for files/other sources which would be good test cases (especially ones with more than one alias required), feel free to push them to the branch.

Also, I couldn't find the test file for the unit registries, so I created a new one.

I haven't added any prose docs/examples to this, any suggestions where such examples would fit in best?

### CODE REVIEW

File: astropy/units/core.py
Comment (pllim) [inline astropy/units/core.py:270]:
```suggestion
        Context manager for temporarily adding aliases for units (e.g., to handle
```

**Review by pllim — COMMENTED** [review]
Thanks! Needs a change log and user facing doc with examples as well.

Making the tests requiring remote data is unnecessary. I propose you store the header in memory as test module variable or something, as all you need for unit parsing is the header object with one `BUNIT` entry (beyond what is mandatory for a simple FITS header). **OR** to make it even simpler, pass in the value of the BUNIT instead of accessing FITS header (e.g., `u.Unit(Angstromsss) == ...`).

**Comment by pllim** [comment]
p.s. You might also want to address codecov comments about lines not covered by tests.

File: astropy/units/core.py
Comment (mhvk) [inline astropy/units/core.py:500]: So, this would become:
```
    # get a context with a new registry, which is a copy of the current one
    context = _UnitContext(get_current_unit_registry())
    # in this new current registry, enable the further units requested
    get_current_unit_registry().add_aliased_units(aliases)
    return context
```

File: astropy/units/tests/test_registry.py
Comment (mhvk) [inline astropy/units/tests/test_registry.py:8]: I do like the very explicit test, but think this one may make more sense in `io.fits` (inside `units`, I'd prefer not to use anything else from astropy).  Aside: if it is from the astropy server, you can use `remote_data(source="astropy")`. Or maybe better to get just the header as @pllim suggests (the name suggests this is a large file!).

I'd then add some very simple tests just using `u.Unit('counts')`, like @pllim suggested.

For the location, you could just use `test_units.py`. That file is where `add_enabled_units` is tested as well - though "tested" is a bit of an overstatement with just one or two tests! And `set_enabled_units` is not tested explicitly at all...

Alternatively, since `test_units.py` is very long and there are rather few registry tests in `test_units.py` anyway, you could move those tests here.

**Review by mhvk — COMMENTED** [review]
That looks nice! Some comments inline, mostly about using the function, not the method, as a context manager.

You pose a very good question about where to document this... Looking at our docs, we have a section about "Unrecognized Units" inside "String Representations of Units", and it just describes what happens, not how to solve it. My sense would be to make that its own section "Dealing with Unrecognized Units" and then have a part about how to define an alias. Ideally, we then cross-reference this from, e.g., the fits pages...  

I fear this will be more work than coding it up...  But it will also be very nice. Indeed, amply nice enough that I'm going to add the "what's new needed" label...

**Comment by mhvk** [comment]
Ah, I didn't have to - @pllim already added the label!

**Comment by mhvk** [comment]
@aragilar - this is quite a nice addition. Will you have time to finish it up?  Feature freeze for 4.3 is coming up...

**Comment by aragilar** [comment]
I've expanded the section on Unrecognized Units, and added a changelog entry. I probably won't have a great amount of time before feature freeze to add more documentation (or other changes), but feel free to push changes.

**Comment by mhvk** [comment]
@aragilar - thanks for the docs! I had another look and fixed the bugs. Another thing I realized is that it would be nicer if the name matched a bit better what the other methods of the registry had, so I changed it to `add_enabled_aliases`. Also, most often one probably does not want to add, but just set some aliases, so I added a `set_enabled_aliases` method as well.

p.s. I shortened your changelog fragment, since I think this needs a what's-new -- at least if I can get it to work when reading FITS files!

**Comment by mhvk** [comment]
@adrn - as I contributed to this PR, it is probably good to have another set of eyes for approval...

File: docs/units/format.rst
Comment (adrn) [inline docs/units/format.rst:200]:
```suggestion
legacy or misspelt unit strings into their standardized form,
```
(for consistency, we use US spelling in the documentation https://docs.astropy.org/en/stable/development/style-guide.html#spelling)

File: docs/units/format.rst
Comment (adrn) [inline docs/units/format.rst:228]:
```suggestion
By adding additional unit aliases for the misspelt units with
```

File: docs/units/format.rst
Comment (adrn) [inline docs/units/format.rst:229]:
```suggestion
`~astropy.units.set_enabled_aliases` (e.g., 'Angstroms' for 'Angstrom') as demonstrated below, and defining new units
```

File: docs/units/format.rst
Comment (adrn) [inline docs/units/format.rst:239]: Hm, I found the phrasing here a little confusing.

File: astropy/units/core.py
Comment (adrn) [inline astropy/units/core.py:277]:
```suggestion
        Set aliases for units.
        
        This is useful for handling alternate spellings of unit names, or misspelled unit names.
```

File: astropy/units/core.py
Comment (adrn) [inline astropy/units/core.py:283]:
```suggestion
            must be the `astropy.units.Unit` that the alias will be mapped to. For example, ``{"parsnek": u.parsec}``.
```
(actual example might be too much of a joke but I think it would be good to have some small example here?

File: astropy/units/core.py
Comment (adrn) [inline astropy/units/core.py:288]: Does this get raised? Maybe a copy-paste error from `add_`?

File: astropy/units/core.py
Comment (adrn) [inline astropy/units/core.py:295]:
```suggestion
        Add aliases for units.
        
        This is useful for handling alternate spellings of unit names, or misspelled unit names.
```

File: astropy/units/core.py
Comment (adrn) [inline astropy/units/core.py:301]: Same example as above?

File: astropy/units/core.py
Comment (adrn) [inline astropy/units/core.py:504]: Hm, I guess same comment as above.

File: astropy/units/core.py
Comment (adrn) [inline astropy/units/core.py:536]: And once more...

**Review by adrn — CHANGES_REQUESTED** [review]
Mostly documentation checks so far. I think the code and tests look fine to me!

File: astropy/units/core.py
Comment (mhvk) [reply astropy/units/core.py:288]: It can get raised if one is trying to define an alias using a string that already exists as a regular unit

Comment (adrn) [reply astropy/units/core.py:288]: Ah, right

File: docs/units/format.rst
Comment (mhvk) [reply docs/units/format.rst:200]: OK, done.

File: docs/units/format.rst
Comment (mhvk) [reply docs/units/format.rst:229]: Rephrased.

File: docs/units/format.rst
Comment (mhvk) [reply docs/units/format.rst:239]: Fortunately, I found a different method that means all formats are now supported, so this can go. Will push a revised PR shortly.

File: astropy/units/core.py
Comment (mhvk) [reply astropy/units/core.py:277]: Done

File: astropy/units/core.py
Comment (mhvk) [reply astropy/units/core.py:504]: I ended up thinking the sentence was mostly useful for the functions, much less so in the method (which the user essentially never sees...)

File: astropy/units/core.py
Comment (mhvk) [reply astropy/units/core.py:283]: I'd tend to just leave it to the more commonly used functions, which I did give an example.

**Comment by mhvk** [comment]
@adrn - I rather changed the implementation, to have the actual format parsers become aware of the aliases. This seemed the only way to ensure it would work also in FITS, CDS, VOUnit, and OGIP. Conceptually, now if the parser does not recognize an individual unit string, it will check the aliases (in all formats; possibly not quite logical, since most do not use the registry, but certainly easiest...).

So, this needs a re-review, I fear... I added some comments in-line.

File: astropy/units/format/ogip.py
Comment (mhvk) [inline astropy/units/format/ogip.py:354]: This is inherited directly from generic, so was superfluous to start with.

File: astropy/units/format/vounit.py
Comment (mhvk) [inline astropy/units/format/vounit.py:105]: Moved from the original `_parse_unit` - so that any custom units are defined *after* possible aliases are checked. Arguably perhaps even a bit more logical.

**Comment by mhvk** [comment]
OK, solved a stupid bug and added some tests inside `io.fits` to show that it also works there.

File: astropy/units/core.py
Comment (adrn) [inline astropy/units/core.py:319]: Removing the word "too" in the 2nd error message, I think these two if clauses can be "or"'d and have just one error to raise? I like the phrasing of the second better ("already is an alias for" vs. "already means")

File: docs/units/format.rst
Comment (adrn) [inline docs/units/format.rst:230]:
```suggestion
demonstrated below), or defining new units via `~astropy.units.def_unit` and
```

File: docs/units/format.rst
Comment (adrn) [inline docs/units/format.rst:243]:
```suggestion
dictionary mapping the misspelt string to an astropy unit. The following code
snippet shows how to set up an alias for a hypothetical nonstandard unit 
"Angstroem" to map to the standard unit Angstrom::
```

File: docs/units/format.rst
Comment (adrn) [inline docs/units/format.rst:261]:
```suggestion
The aliases can be reset by passing an empty dictionary::
```

**Review by adrn — APPROVED** [review]
New implementation looks good! Some suggested documentation changes.

File: astropy/units/core.py
Comment (mhvk) [reply astropy/units/core.py:319]: I had that initially but then the f-string becomes annoying, since one has to switch between `self._registry` and `self._aliases`...  I'll just delete "too"...

File: docs/units/format.rst
Comment (mhvk) [reply docs/units/format.rst:230]: Done

File: docs/units/format.rst
Comment (mhvk) [reply docs/units/format.rst:261]: Ha! That's an odd one. Fixed.

**Comment by mhvk** [comment]
@adrn - thanks for the review. Since it is all docs, I'll push with `[skip ci]`

**Comment by mhvk** [comment]
@aragilar - thanks for starting this, and apologies for running away with your PR...  But it will now be in 4.3!

**Comment by adrn** [comment]
Failure unrelated, and previously approved

**Comment by aragilar** [comment]
@mhvk That's cool, there's even support for the other formats, which is even better!

**Pull request merged**
